Mathlib/Logic/Function/Basic.lean

```diff
@@ -135,6 +136,11 @@ theorem Injective.piMap {ι : Sort*} {α β : ι → Sort*} {f : ∀ i, α i →
 theorem Injective.comp_left {g : β → γ} (hg : Injective g) : Injective (g ∘ · : (α → β) → α → γ) :=
   .piMap fun _ ↦ hg
 
+theorem injective_comp_left_iff [Nonempty α] {g : β → γ} :
+    Injective (g ∘ · : (α → β) → α → γ) ↔ Injective g :=
+  ⟨fun h b₁ b₂ eq ↦ congr_fun (h (a₁ := fun _ ↦ b₁) (a₂ := fun _ ↦ b₂) <| funext fun _ ↦ eq)
+    (Classical.arbitrary α), (·.comp_left)⟩
```

Comment:
This is maybe close enough to foundations that the axiom of choice will offend some people. Can you use the recursion on Nonempty instead, either with `let` or `inhabit`?

Reply:
(and throughout)

Reply:
I switched to using `Nonempty.elim` here and eliminated the use of `Classical.choice` in this theorem. There's one other `Classical.arbitrary` in `surjective_comp_left_iff`, but I'm not changing it since I don't think the right-to-left direction can be proven without using choice anyway.